mlx/io/gguf.cpp

```diff
@@ -70,6 +79,72 @@ std::tuple<allocator::Buffer, Dtype> extract_tensor_data(gguf_tensor* tensor) {
   return {buffer, float16};
 }
 
+void extract_q4_1_data(
+    std::unordered_map<std::string, array>* a,
+    const gguf_tensor& tensor) {
+  std::string name = std::string(tensor.name, tensor.namelen);
+  const std::vector<int> shape = get_shape(tensor);
+  const uint64_t weights_per_byte = 2;
+  const uint64_t weights_per_block = 32;
+  if (shape[shape.size() - 1] % weights_per_block != 0) {
+    std::ostringstream msg;
+    msg << "[save_gguf] tensor " << name
+        << "has incompatible last dim shape: " << shape[shape.size() - 1];
+    throw std::runtime_error(msg.str());
+  }
+  const uint64_t bytes_per_block =
+      20; // 2 bytes scale, 2 bytes bias, 32x0.5 byte weights
+  const uint64_t num_blocks = tensor.num_weights / weights_per_block;
+  allocator::Buffer weights_buffer =
+      allocator::malloc(tensor.num_weights / weights_per_byte);
+  allocator::Buffer scales_buffer = allocator::malloc(num_blocks * 2);
+  allocator::Buffer biases_buffer = allocator::malloc(num_blocks * 2);
+  auto data = (uint8_t*)tensor.weights_data;
+  auto weigths = (uint8_t*)weights_buffer.raw_ptr();
+  auto scales = (uint16_t*)scales_buffer.raw_ptr();
+  auto biases = (uint16_t*)biases_buffer.raw_ptr();
+  for (int64_t i = 0; i < num_blocks; i++) {
+    uint8_t* block_data = data + i * bytes_per_block;
+    scales[i] = *((uint16_t*)block_data);
+    biases[i] = *((uint16_t*)block_data + 1);
+    // 16 weights in lower bits
+    for (int64_t j = 0; j < 16; ++j) {
+      uint8_t x = (block_data[j + 4] & 0x0F);
+      if (j % 2 == 0) {
+        x <<= 4;
+      }
+      weigths[i * 16 + j / 2] += x;
+    }
+    // 16 weights in higher bits
+    for (int64_t j = 0; j < 16; ++j) {
+      uint8_t x = (block_data[j + 4] >> 4);
+      if (j % 2 == 0) {
+        x <<= 4;
+      }
+      weigths[i * 16 + 8 + j / 2] += x;
+    }
+  }
+  std::vector<int> weights_shape = shape;
+  weights_shape[shape.size() - 1] =
+      weights_shape[shape.size() - 1] / weights_per_byte / 4;
+  a->insert({name, array(weights_buffer, weights_shape, uint32)});
+
+  const std::string weight_suffix = ".weight";
```

Comment:
@jbochi is there a way to tell from the tensor meta data if it is quantized? Hardcoding the `weight` name does not seem like a good idea.

Reply:
Not at the moment. I considered adding a separate tensor for it. We can infer the block size by dividing the number of weights by the number of scales or biases, but we can't be sure about the number of bits per weight.

We could also add return a second map with all the metadata